Recruit-Training/Advanced-Recruit-Training/Telemetry-Teaser/src/App.tsx

```diff
@@ -1,25 +1,113 @@
-    return;
-  }
+    event?.preventDefault();
+    setVisible(true);
+  };
+
+  const BatteryChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = parseInt(event.target.value);
+    setVisible(false);
+    if (value) {
+      setInputs({ ...inputs, battery: value });
+    } else {
+      setInputs({ ...inputs, battery: undefined });
+    }
+  };
+
+  const SpeedChange = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const value = parseInt(event.target.value);
+    setVisible(false);
+    if (value !== undefined) {
+      setInputs({ ...inputs, speed: value });
+    } else {
+      setInputs({ ...inputs, speed: undefined });
+    }
+  };
+
+  const sliderChange = (num: string) => {
+    const value = parseFloat(num);
+    setInputs({ ...inputs, weather: value });
+    setVisible(false);
+  };
 
   return (
     <div className="h-screen w-screen bg-[#212121]">
       <div className="flex h-full flex-col items-center pt-36 text-white">
         <Header />
         <form name="simulator" className="flex w-full flex-col items-center">
           <div className="mb-4 flex w-full flex-col items-center gap-y-4">
-            <SpeedInput />
-            <BatteryInput />
+            <SpeedInput value={inputs["speed"]} onChange={SpeedChange} />
+            {inputs["speed"] !== undefined &&
+              (inputs["speed"] > 100 || inputs["speed"] < 0) && (
+                <p className="text-red-500">
+                  {" "}
+                  The speed should be within the range of 0 to 90
+                </p>
+              )}
+            {inputs["speed"] === undefined && (
+              <p className="text-red-500">Speed is required</p>
+            )}
+            <BatteryInput value={inputs["battery"]} onChange={BatteryChange} />
+            {inputs["battery"] !== undefined &&
+              (inputs["battery"] > 100 || inputs["battery"] < 0) && (
+                <p className="text-red-500">
+                  The battery percentage should be with the range of 0 to 100
+                </p>
+              )}
+            {inputs["battery"] === undefined && (
+              <p className="text-red-500">Battery percentage is required</p>
+            )}
           </div>
           <div className="flex w-full flex-row justify-center gap-4">
-            <WeatherInput />
+            <WeatherInput value={inputs["weather"]} onChange={sliderChange} />
           </div>
+
+          <button
+            className="mt-10 h-10 w-60 rounded bg-blue-200 text-white"
+            onClick={() => calculateRange()}
+          >
+            Submit
+          </button>
+          {visible &&
+            inputs["speed"] !== undefined &&
+            inputs["battery"] !== undefined &&
+            inputs["weather"] !== undefined &&
+            inputs["battery"] <= 100 &&
+            inputs["battery"] >= 0 &&
+            inputs["speed"] <= 100 &&
+            inputs["speed"] >= 0 && (
+              <p className="mt-10">
```

Comment:
you're doing a lot of recomputation here, can you derive the valid and invalid states of each input field and reuse that variable?